dali/dfu/dfurun.cpp

```diff
@@ -1355,23 +1341,24 @@ class CDFUengine: public CInterface, implements IDFUengine
                         }
 
                         bool dirPerPart = false;
-                        //MORE: This could be combined with the code that gets defaultSprayParts
-                        if (isContainerized())
+                        IDFUfileSpec *dst = wu->queryUpdateDestination();
+                        StringBuffer clusterName;
+                        destination->getGroupName(0, clusterName);
+                        Owned<IPropertyTree> plane = getStoragePlane(clusterName);
+                        if (isContainerized() && plane)
+                            dirPerPart = plane->getPropBool("@subDirPerFilePart", true);
+                        // keys default wrap for copy
+                        if (destination->getWrap()||(iskey&&(cmd==DFUcmd_copy)))
                         {
-                            StringBuffer clusterName;
-                            destination->getGroupName(0, clusterName);
-                            Owned<IPropertyTree> plane = getStoragePlane(clusterName);
-                            if (plane)
-                            {
-                                dirPerPart = plane->getPropBool("@subDirPerFilePart", true);
-                            }
+                            if (destination->getNumPartsOverride())
+                                throw makeStringExceptionV(-1, "DestinationNumPartOverride is provided but %s", destination->getWrap()?"getWrap is true":"is copying a key");
+                            dst->setNumPartsOverride(srcFile->numParts());
                         }
-
-                        if (destination->getWrap())
+                        else if (isContainerized() && plane)
```

Comment:
Could probably be true for non-containerized as well.  (If we are to eventually support remote non-containerized storage cleanly.)

Reply:
I had it like this before.  However, @jakesmith suggested it may cause problems in mixed environments: https://github.com/hpcc-systems/HPCC-Platform/pull/17192#discussion_r1159777220

@jakesmith

Reply:
https://github.com/hpcc-systems/HPCC-Platform/pull/17192#discussion_r1159777220 was about not setting subDirPerPart in non-containerized..

the code being commented on here, i.e.:
```
                        else if (isContainerized() && plane)
                        {
                            // use destination defaultSprayParts if requestor doesn't provide num parts
                            if (plane->hasProp("@defaultSprayParts") && destination->getNumPartsOverride()==0)
                                dst->setNumPartsOverride(plane->getPropInt("@defaultSprayParts"));
                        }
```

afaics, that could equally apply to non-containerized, so yes, I'd make the condition here if (plane) only.